---
title: "Advanced plotting with schnapps"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Advanced plotting with schnapps}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  warning=FALSE, message=FALSE, 
  comment = "#>"
)
```

```{r setup}
library(schnapps)
```

## Background
This vignette will illustrate some advanced plotting functionality available in `schnapps` that allows plotting single cell profiles with ideograms and gene annotations. This additional functionaility requires the following bioconductor packages (which are not installed automatically): `regioneR`, `karyploteR`, `biomaRt` and `GenomeInfoDb`.

```{r}
library(regioneR)
library(karyoploteR)
library(biomaRt)
library(GenomeInfoDb)
```

## Data processing
We'll now process the data as in the introductory vignette and call haplotype specific copy number.
```{r}
data("haplotypes")
data("CNbins")
haplotypes <- format_haplotypes_dlp(haplotypes, CNbins)
hscn <- callHaplotypeSpecificCN(CNbins, haplotypes, phasebyarm = FALSE)
```

## Plotting
We'll now run through a series of examples that makes use of this plotting functionality.

### Genome wide plot with ideogram
```{r, fig.width=15}
plotideogram(hscn, cellid = "SA922-A90554B-R35-C61")
```

### Genome wide plot with ideogram and gene annotations
```{r, fig.width=15}
plotideogram(hscn, cellid = "SA922-A90554B-R35-C61", gene.symbols = c("TP53", "MYC"))
```

### Chromosome level plots with ideogram and gene annotations
```{r, fig.width = 10}
plotideogram(hscn, cellid = "SA922-A90554B-R31-C42", chr = c("8"), gene.symbols = c("MYC"))
```

```{r, fig.width = 10}
plotideogram(hscn, cellid = "SA922-A90554B-R34-C67", chr = c("17"), gene.symbols = c("TP53"))
```



