<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Advanced plotting with schnapps • schnapps</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Advanced plotting with schnapps">
<meta property="og:description" content="schnapps">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">schnapps</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ASCN.html">Allele Specific Copy Number Inference</a>
    </li>
    <li>
      <a href="../articles/BetaBinom.html">Beta-Binomial model for schnapps</a>
    </li>
    <li>
      <a href="../articles/karyoplot.html">Advanced plotting with schnapps</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="karyoplot_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Advanced plotting with schnapps</h1>
            
      
      
      <div class="hidden name"><code>karyoplot.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">schnapps</span>)
</pre></div>
<div id="background" class="section level2">
<h2 class="hasAnchor">
<a href="#background" class="anchor"></a>Background</h2>
<p>This vignette will illustrate some advanced plotting functionality available in <code>schnapps</code> that allows plotting single cell profiles with ideograms and gene annotations. This additional functionaility requires the following bioconductor packages (which are not installed automatically): <code>regioneR</code>, <code>karyploteR</code>, <code>biomaRt</code> and <code>GenomeInfoDb</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">regioneR</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw"><a href="https://github.com/bernatgel/karyoploteR">karyoploteR</a></span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">biomaRt</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="kw">GenomeInfoDb</span>)
</pre></div>
</div>
<div id="data-processing" class="section level2">
<h2 class="hasAnchor">
<a href="#data-processing" class="anchor"></a>Data processing</h2>
<p>We’ll now process the data as in the introductory vignette and call haplotype specific copy number.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"haplotypes"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"CNbins"</span>)
<span class="kw">haplotypes</span> <span class="op">&lt;-</span> <span class="fu">format_haplotypes_dlp</span>(<span class="kw">haplotypes</span>, <span class="kw">CNbins</span>)
<span class="kw">hscn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/callHaplotypeSpecificCN.html">callHaplotypeSpecificCN</a></span>(<span class="kw">CNbins</span>, <span class="kw">haplotypes</span>, phasebyarm = <span class="fl">FALSE</span>)
</pre></div>
</div>
<div id="plotting" class="section level2">
<h2 class="hasAnchor">
<a href="#plotting" class="anchor"></a>Plotting</h2>
<p>We’ll now run through a series of examples that makes use of this plotting functionality.</p>
<div id="genome-wide-plot-with-ideogram" class="section level3">
<h3 class="hasAnchor">
<a href="#genome-wide-plot-with-ideogram" class="anchor"></a>Genome wide plot with ideogram</h3>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="fu">plotideogram</span>(<span class="kw">hscn</span>, cellid = <span class="st">"SA921-A90554A-R05-C07"</span>)
</pre></div>
<p><img src="karyoplot_files/figure-html/unnamed-chunk-4-1.png" width="1440"></p>
<pre><code>#&gt; expression(kp &lt;- karyoploteR::plotKaryotype(plot.type = 4, plot.params = pp, 
#&gt;     chromosomes = chr), karyoploteR::kpAxis(kp, r0 = 0.52, r1 = 1), 
#&gt;     karyoploteR::kpPoints(kp, data = data, y = data$BAF, r0 = 0.52, 
#&gt;         r1 = 1, col = data$col_ASstate), karyoploteR::kpAxis(kp, 
#&gt;         tick.pos = seq(0, maxCN, 2), r0 = 0, r1 = 0.48, ymax = maxCN, 
#&gt;         ymin = 0), karyoploteR::kpPoints(kp, data = data, y = data$copy, 
#&gt;         r0 = 0, r1 = 0.48, ymin = 0, ymax = maxCN, col = data$col_state))</code></pre>
</div>
<div id="genome-wide-plot-with-ideogram-and-gene-annotations" class="section level3">
<h3 class="hasAnchor">
<a href="#genome-wide-plot-with-ideogram-and-gene-annotations" class="anchor"></a>Genome wide plot with ideogram and gene annotations</h3>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="fu">plotideogram</span>(<span class="kw">hscn</span>, cellid = <span class="st">"SA921-A90554A-R05-C07"</span>, gene.symbols = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"TP53"</span>, <span class="st">"MYC"</span>))
</pre></div>
<p><img src="karyoplot_files/figure-html/unnamed-chunk-5-1.png" width="1440"></p>
<pre><code>#&gt; expression(kp &lt;- karyoploteR::plotKaryotype(plot.type = 4, plot.params = pp, 
#&gt;     chromosomes = chr), karyoploteR::kpAxis(kp, r0 = 0.4, r1 = 0.75), 
#&gt;     karyoploteR::kpPoints(kp, data = data, y = data$BAF, r0 = 0.4, 
#&gt;         r1 = 0.75, col = data$col_ASstate), karyoploteR::kpAxis(kp, 
#&gt;         tick.pos = seq(0, maxCN, 2), r0 = 0, r1 = 0.35, ymax = maxCN, 
#&gt;         ymin = 0), karyoploteR::kpPoints(kp, data = data, y = data$copy, 
#&gt;         r0 = 0, r1 = 0.35, ymin = 0, ymax = maxCN, col = data$col_state), 
#&gt;     karyoploteR::kpPlotMarkers(kp, data = genes, labels = genes$hgnc_symbol, 
#&gt;         line.color = "#555555", marker.parts = c(0.95, 0.025, 
#&gt;             0.025), r1 = 1.05))</code></pre>
</div>
<div id="chrosome-level-plots-with-ideogram-and-gene-annotations" class="section level3">
<h3 class="hasAnchor">
<a href="#chrosome-level-plots-with-ideogram-and-gene-annotations" class="anchor"></a>Chrosome level plots with ideogram and gene annotations</h3>
<div class="sourceCode" id="cb8"><pre class="downlit">
<span class="fu">plotideogram</span>(<span class="kw">hscn</span>, cellid = <span class="st">"SA921-A90554A-R05-C07"</span>, chr = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"8"</span>), gene.symbols = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"MYC"</span>))
</pre></div>
<p><img src="karyoplot_files/figure-html/unnamed-chunk-6-1.png" width="960"></p>
<pre><code>#&gt; expression(kp &lt;- karyoploteR::plotKaryotype(plot.type = 4, plot.params = pp, 
#&gt;     chromosomes = chr), karyoploteR::kpAxis(kp, r0 = 0.4, r1 = 0.75), 
#&gt;     karyoploteR::kpPoints(kp, data = data, y = data$BAF, r0 = 0.4, 
#&gt;         r1 = 0.75, col = data$col_ASstate), karyoploteR::kpAxis(kp, 
#&gt;         tick.pos = seq(0, maxCN, 2), r0 = 0, r1 = 0.35, ymax = maxCN, 
#&gt;         ymin = 0), karyoploteR::kpPoints(kp, data = data, y = data$copy, 
#&gt;         r0 = 0, r1 = 0.35, ymin = 0, ymax = maxCN, col = data$col_state), 
#&gt;     karyoploteR::kpPlotMarkers(kp, data = genes, labels = genes$hgnc_symbol, 
#&gt;         line.color = "#555555", marker.parts = c(0.95, 0.025, 
#&gt;             0.025), r1 = 1.05))</code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="fu">plotideogram</span>(<span class="kw">hscn</span>, cellid = <span class="st">"SA921-A90554A-R05-C07"</span>, chr = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"17"</span>), gene.symbols = <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"TP53"</span>))
</pre></div>
<p><img src="karyoplot_files/figure-html/unnamed-chunk-7-1.png" width="960"></p>
<pre><code>#&gt; expression(kp &lt;- karyoploteR::plotKaryotype(plot.type = 4, plot.params = pp, 
#&gt;     chromosomes = chr), karyoploteR::kpAxis(kp, r0 = 0.4, r1 = 0.75), 
#&gt;     karyoploteR::kpPoints(kp, data = data, y = data$BAF, r0 = 0.4, 
#&gt;         r1 = 0.75, col = data$col_ASstate), karyoploteR::kpAxis(kp, 
#&gt;         tick.pos = seq(0, maxCN, 2), r0 = 0, r1 = 0.35, ymax = maxCN, 
#&gt;         ymin = 0), karyoploteR::kpPoints(kp, data = data, y = data$copy, 
#&gt;         r0 = 0, r1 = 0.35, ymin = 0, ymax = maxCN, col = data$col_state), 
#&gt;     karyoploteR::kpPlotMarkers(kp, data = genes, labels = genes$hgnc_symbol, 
#&gt;         line.color = "#555555", marker.parts = c(0.95, 0.025, 
#&gt;             0.025), r1 = 1.05))</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Marc J Williams.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
